CDCgov/prime-reportstream
# [9302] Host RS API Docs Swagger UI from Azure Storage Server

This PR:

Instrumented the Report Stream platform local setup docker compose such that swagger UI for API Docs are generated (from swagger annotated kotlin source files), assembled (spec yaml files and swagger UI resources are copied to build location), and uploaded (az storage container commands) to the azurite server)

More work on authorization part is tracked by issue 10706: 
https://github.com/CDCgov/prime-reportstream/issues/10706

Note: 

1. Per team hosting the API docs swagger ui from azure storage is preferred (vs, #10262 where swagger ui is served as an azure functions end point, there are pros and cons between the approaches, but hosting from a azure storage seems to have scaling advantages)
2. For local setup: CORS settings in start_func.sh are modified to allow azurite server (on port 10000)
3. For deployment to cloud ENVs: DevOp CORS configuration changes are needed, will open DevOp ticket accordingly

Test Steps (local setup):

1. Although in this approach, no kotlin code changes made, it is good still to pass unit tests, integration tests, and smoke tests - following README on how to spin up a local RS server
2. After step 1, RS server is up and running as usual, plus that swagger UI for RS API docs is provisioned at URL:
4. API Docs swagger UI is at: http://127.0.0.1:10000/devstoreaccount1/apidocs/index.html
5. For now, the API docs contains API calls from ApiKeysFunctions.kt:
6. POST: create API key given org, kid, scope
7. GET: retrieve API key given org
8. DELETE: delete API key given org, kid, scope

On apple silicon machine:

1. check out PR into local repo
2. Follow README "Getting started as Dev": https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/getting-started/getting-started.md
3. On apple silicon machine, the azure functions docker won't start, other docker compose services are up and running
4. Run ./gradlew run, this will spin up an azure functions server from native OS, task "run" is instrumented to do the swagger UI upload to the azurite storage server
5. Check the swagger UI contents e.g. using a azure storage explorer - under container "apidocs"
6. Select index.html and right click and select "Copy URL"
7. Point browser to the blob container URL from step 6
8. Should be able to see swagger UI with RS ApiKeysFunctions end points: POST - create key, GET - retrieve key, DELETE - remove key


Some sample screen shots:

1. After a local RS server is up and running, the swagger ui + API specs yaml files should be provisioned in azurite storage server under container "apidocs", use azurite explorer to visually check it out:
10. 
![azurite-explorer-view](https://github.com/CDCgov/prime-reportstream/assets/135686833/71642008-53f4-4afb-a900-6b3cf1712b7b)
11. As shown in step 2, right click the index.html and Copy URL, then paste URL into your browser, this will bring up the swagger UI for RS API:
12. 
 
![swagger-ui-apidocs-page](https://github.com/CDCgov/prime-reportstream/assets/135686833/047931b4-fa78-46da-9f48-bad7bb01172d)

13. Call POST, GET, DELETE to create, retrieve, and delete the public key for an org, as usually do with postman

## Changes
- *Include a comprehensive list of changes in this PR*
- *(For web UI changes) Include screenshots/video of changes*

## Checklist

### Testing

### Process

## Linked Issues
- #9302
- #10262

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #10706 
- #10224

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
